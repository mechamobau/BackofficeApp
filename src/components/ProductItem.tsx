import React from 'react';
import {StyleSheet, Text, View} from 'react-native';
import useTheme from '../hooks/useTheme';
import Product from '../models/Product';
import RoundedControlButton from './RoundedControlButton';

type BaseProps = Pick<Product, 'name' | 'quantity' | 'value'>;

type Props = BaseProps & {
  onRemovePress?: () => void;
  onAddPress?: () => void;
};

function ProductItem(props: Props) {
  const {name, quantity, value, onRemovePress, onAddPress} = props;

  const formattedValue = value.toFixed(2);

  const {theme} = useTheme();

  return (
    <View style={styles.container} accessible>
      <View>
        <Text style={[styles.title, {color: theme.textColor}]}>{name}</Text>
        <Text style={[styles.value, {color: theme.textColor}]}>
          ${formattedValue}
        </Text>
      </View>
      <View style={styles.quantityControlWrapper}>
        {quantity >= 1 && (
          <RoundedControlButton
            accessibilityRole="button"
            accessibilityLabel="Remover produto"
            onPress={onRemovePress}
          >
            -
          </RoundedControlButton>
        )}
        <Text
          style={[styles.quantity, {color: theme.textColor}]}
          accessibilityLabel={`NÃºmero de produtos: ${quantity}`}
        >
          {quantity}
        </Text>
        <RoundedControlButton
          accessibilityRole="button"
          accessibilityLabel="Adicionar produto"
          onPress={onAddPress}
        >
          +
        </RoundedControlButton>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  title: {
    fontWeight: 'bold',
    fontSize: 24,
    marginBottom: 6,
  },
  value: {
    fontSize: 16,
  },
  quantity: {
    marginVertical: 0,
    marginHorizontal: 15,
    fontSize: 19,
  },
  mainInfoWrapper: {},
  quantityControlWrapper: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    paddingVertical: 12,
    paddingHorizontal: 10,
    marginBottom: 12,
    width: '100%',
  },
});

export default ProductItem;
