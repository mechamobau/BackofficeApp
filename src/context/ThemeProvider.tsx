import React from 'react';
import type {ReactNode} from 'react';

import COLORS from '../constants/COLORS';
import {useColorScheme} from 'react-native';

export type ThemeType = typeof lightTheme; // This is the type definition for my theme object.

export const lightTheme = {
  textColor: COLORS.black,
  placeholderColor: COLORS.dark,
  buttonBorderColor: COLORS.black,
};

export const darkTheme: ThemeType = {
  textColor: COLORS.white,
  placeholderColor: COLORS.light,
  buttonBorderColor: COLORS.lighter,
};

type Theme = {
  theme: ThemeType;
  isDark: boolean;
};

export const ThemeContext = React.createContext<Theme>({
  theme: lightTheme,
  isDark: false,
});

type Props = {
  children: ReactNode;
};

function ThemeProvider({children}: Props) {
  const colorScheme = useColorScheme();

  const isDark = colorScheme === 'dark';

  const value = {
    theme: isDark ? darkTheme : lightTheme,
    isDark,
  };

  return (
    <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>
  );
}

export default ThemeProvider;
