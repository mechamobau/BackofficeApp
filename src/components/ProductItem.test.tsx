import React, {ReactNode} from 'react';
import {fireEvent, render} from '@testing-library/react-native';
import ProductItem from './ProductItem';
import ThemeProvider from '../context/ThemeProvider';

const AllProviders = ({children}: {children: ReactNode}) => (
  <ThemeProvider>{children}</ThemeProvider>
);

describe('ProductItem | component | integration test', () => {
  it('renders correctly', () => {
    const {queryByText, queryByA11yLabel} = render(
      <ProductItem name="Product name display" quantity={2} value={100} />,
      {wrapper: AllProviders},
    );

    expect(queryByText('Product name display')).not.toBeNull();
    expect(queryByText('$100.00')).not.toBeNull();
    expect(queryByA11yLabel('NÃºmero de produtos: 2')).not.toBeNull();
  });

  it('calls add and remove quantity handlers on pressing', () => {
    const addFn = jest.fn();
    const removeFn = jest.fn();

    const {getByA11yLabel} = render(
      <ProductItem
        name="Product name display"
        quantity={2}
        value={100}
        onAddPress={addFn}
        onRemovePress={removeFn}
      />,
      {wrapper: AllProviders},
    );

    fireEvent.press(getByA11yLabel('Adicionar produto'));

    expect(addFn).toHaveBeenCalled();

    fireEvent.press(getByA11yLabel('Remover produto'));

    expect(removeFn).toHaveBeenCalled();
  });

  it('matches snapshot', () => {
    const {toJSON} = render(
      <ProductItem name="Product name display" quantity={2} value={100} />,
      {wrapper: AllProviders},
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
