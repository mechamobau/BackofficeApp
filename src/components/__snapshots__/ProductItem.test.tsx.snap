// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem | component | integration test matches snapshot 1`] = `
<View
  accessible={true}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginBottom": 12,
      "paddingHorizontal": 10,
      "paddingVertical": 12,
      "width": "100%",
    }
  }
>
  <View>
    <Text
      style={
        Array [
          Object {
            "fontSize": 18,
            "fontWeight": "bold",
            "marginBottom": 6,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      Product name di...
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontSize": 16,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      Valor unitário: $
      100.00
    </Text>
    <Text
      style={
        Array [
          Object {
            "fontSize": 16,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      Valor total: $
      200.00
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      accessibilityLabel="Diminuir quantidade"
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#000000",
          "borderRadius": 20,
          "borderWidth": 1,
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 40,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 22,
            },
            Object {
              "color": "#000000",
            },
          ]
        }
      >
        -
      </Text>
    </View>
    <Text
      accessibilityLabel="Número de produtos: 2"
      style={
        Array [
          Object {
            "fontSize": 19,
            "marginHorizontal": 15,
            "marginVertical": 0,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      2
    </Text>
    <View
      accessibilityLabel="Aumentar quantidade"
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#000000",
          "borderRadius": 20,
          "borderWidth": 1,
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 40,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 22,
            },
            Object {
              "color": "#000000",
            },
          ]
        }
      >
        +
      </Text>
    </View>
    <View
      accessibilityLabel="Remover produto"
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "red",
          "borderRadius": 20,
          "borderWidth": 1,
          "height": 40,
          "justifyContent": "center",
          "marginLeft": 15,
          "opacity": 1,
          "width": 40,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 22,
            },
            Object {
              "color": "red",
            },
          ]
        }
      >
        X
      </Text>
    </View>
  </View>
</View>
`;
